acvm_backend_barretenberg/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+#![warn(unused_crate_dependencies, unused_extern_crates)]
+#![warn(unreachable_pub)]
+
+// `acvm-backend-barretenberg` can either interact with the Barretenberg backend through a static library
+// or through an embedded wasm binary. It does not make sense to include both of these backends at the same time.
+// We then throw a compilation error if both flags are set.
+// TODO: handle JS target.
+#[cfg(all(feature = "native", feature = "wasm"))]
+compile_error!("feature \"native\"  and feature \"wasm\" cannot be enabled at the same time");
+
+mod acvm_interop;
+mod blake2s;
+mod composer;
+#[cfg(all(feature = "native", test))]
+mod crs;
+mod pedersen;
+mod pippenger;
+mod scalar_mul;
+mod schnorr;
+
+/// The number of bytes necessary to store a `FieldElement`.
+const FIELD_BYTES: usize = 32;
+
+pub struct Barretenberg {
+    #[cfg(feature = "wasm")]
+    memory: Memory,
+    #[cfg(feature = "wasm")]
+    instance: Instance,
+}
+
+impl Default for Barretenberg {
+    fn default() -> Barretenberg {
+        Barretenberg::new()
+    }
+}
+
+#[test]
+fn smoke() {
+    use crate::pedersen::Pedersen;
+
+    let b = Barretenberg::new();
+    let (x, y) = b.encrypt(vec![
+        common::acvm::FieldElement::zero(),
+        common::acvm::FieldElement::one(),
+    ]);
+    dbg!(x.to_hex(), y.to_hex());
+}
+
+cfg_if::cfg_if! {
+    if #[cfg(feature = "native")] {
+
+        impl Barretenberg {
+            pub(crate) fn new() -> Barretenberg {
+                Barretenberg {}
+            }
+        }
+
+        fn field_to_array(f: &common::acvm::FieldElement) -> [u8; 32] {
```

Comment:
I'm surprised this wouldn't be used in wasm, are we hand-rolling it or something?

Reply:
This is due to the fact that we need to pass `[u8;32]` to barretenberg_sys but we can just pass `&[u8]` when writing to wasm memory.